sequencer/src/main/java/net/consensys/linea/rpc/methods/LineaSendBundle.java

```diff
@@ -80,32 +90,57 @@ public BundleResponse execute(final PluginRpcRequest request) {
                           .reduce(Bytes::concatenate)
                           .map(Hash::hash));
 
-      if (optBundleHash.isPresent()) {
-        Hash bundleHash = optBundleHash.get();
-        final List<Transaction> txs =
-            bundleParams.txs.stream().map(DomainObjectDecodeUtils::decodeRawTransaction).toList();
-
-        if (!txs.isEmpty()) {
-          bundlePool.putOrReplace(
-              bundleHash,
-              new TransactionBundle(
-                  bundleHash,
-                  txs,
-                  bundleParams.blockNumber,
-                  bundleParams.minTimestamp,
-                  bundleParams.maxTimestamp,
-                  bundleParams.revertingTxHashes()));
-          return new BundleResponse(bundleHash.toHexString());
-        }
-      }
-      // otherwise boom.
-      throw new RuntimeException("Malformed bundle, no bundle transactions present");
+      return optBundleHash
+          .map(
+              bundleHash -> {
+                final List<Transaction> txs =
+                    bundleParams.txs.stream()
+                        .map(DomainObjectDecodeUtils::decodeRawTransaction)
+                        .toList();
+
+                bundlePool.putOrReplace(
+                    bundleHash,
+                    new TransactionBundle(
+                        bundleHash,
+                        txs,
+                        bundleParams.blockNumber,
+                        bundleParams.minTimestamp,
+                        bundleParams.maxTimestamp,
+                        bundleParams.revertingTxHashes()));
+                return new BundleResponse(bundleHash.toHexString());
+              })
+          .orElseThrow(
+              () ->
+                  // otherwise boom.
+                  new RuntimeException("Malformed bundle, no bundle transactions present"));
 
     } catch (final Exception e) {
       throw new PluginRpcEndpointException(new LineaSendBundleError(e.getMessage()));
     }
   }
 
+  private void parametersValidation(final BundleParameter bundleParams) {
+    // synchronized to avoid that 2 parallel requests with the same parameters
+    // will be both processed
+    synchronized (recentlySeenRequestsCache) {
+      final var alreadySeenAt = recentlySeenRequestsCache.getIfPresent(bundleParams);
+      if (alreadySeenAt != null) {
+        throw new IllegalArgumentException(
+            "request already seen " + Duration.between(alreadySeenAt, Instant.now()) + " ago");
+      }
+      recentlySeenRequestsCache.put(bundleParams, Instant.now());
+    }
+
+    if (bundleParams.blockNumber <= blockchainService.getChainHeadHeader().getNumber()) {
+      throw new IllegalArgumentException("bundle block number already in chain");
```

Comment:
worth including bundle block number in the exception message?

Reply:
done